### PR TITLE
Add a way to format text with markers

### PR DESCRIPTION
**Pull request by jorolf**
naming probably needs some improvement

### CODE REVIEW

File: osu.Framework.Desktop.Tests/Visual/TestCaseFormattedText.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                if (markers.Contains(Markers.Red))
+                    text.Colour = Color4.Red;
+
+                if (markers.Contains(Markers.Small))
+                    text.TextSize = 15;
+            }
+        }
+
+        private class MarkdownContainerEndless : MarkdownContainer
```
Comment (peppy) [inline osu.Framework.Desktop.Tests/Visual/TestCaseFormattedText.cs]: why would this ever be required?

File: osu.Framework/Graphics/Containers/FormattedTextFlowContainer.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+using System.Collections.Generic;
+using System.Text;
+using System.Linq;
+using osu.Framework.Graphics.Sprites;
+using System;
+
+namespace osu.Framework.Graphics.Containers
+{
+    public abstract class FormattedTextFlowContainer<T> : TextFlowContainer
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/FormattedTextFlowContainer.cs]: documentation, what is `T`?

File: osu.Framework.Desktop.Tests/Visual/TestCaseFormattedText.cs (hunk 1 above)
Comment (jorolf) [inline osu.Framework.Desktop.Tests/Visual/TestCaseFormattedText.cs]: ~~Minecraft does it that way.~~ I don't know.

File: osu.Framework/Graphics/Containers/FormattedTextFlowContainer.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    {
+        /// <summary>
+        /// Contains strings that tell the parser what a marker is
+        /// </summary>
+        protected abstract Dictionary<string, T> MarkerDelimeters { get; }
+
+        /// <summary>
+        /// If a maker needs both a start and an end or if it should continue until the end if not closed
```
Comment (RemieRichards) [inline osu.Framework/Graphics/Containers/FormattedTextFlowContainer.cs]: maker -> marker

**Comment by Tom94** [comment]
Would be fair to mention that the parsing logic is something I wrote and pushed into an old PR on the osu-side.

**Comment by peppy** [comment]
If this is going to go anywhere it needs unit tests for edge cases, such as unbalanced closures or incorrectly ordered closures.

**Comment by jorolf** [comment]
what do you mean by unbalanced closures or incorrectly ordered closures?

**Comment by peppy** [comment]
`****unbalanced***`
`*_incorrectly ordered*_`

**Comment by default0** [comment]
Others that might be of note:
``_*_`` <- should the * itself be formatted given that it cant match anywhere? If yes, then how about this:
``_*_*`` would the * get formatted? or the _? or is this an error?
``* do I get the format of the asterisk? * or do I? * or is this an error?``
``how does *this
work*
across _multiple
lines?_``

Making things be errors may be annoying for usability unless we offer a method for sanitizing inputs (as you might want to use this in conjunction with untrusted inputs), which is also of note.

**Comment by peppy** [comment]
We should make a decision whether this should be merged or closed. I'd argue that it current falls short of being useful as it isn't a full markdown parser, which is the only use-case i can think of.

**Comment by peppy** [comment]
Okay so we need to make a decision on whether this should be merged or not. It seems like a sound implementation for what it does, but for osu! purposes we will likely need something more full-blown to handle markdown.

Can we get a show or 👍  or 👎  as to whether people think this should exist in the framework or if it is unnecessary?

@Tom94 @default0 @smoogipoo @ppy/contributors

**Comment by smoogipoo** [comment]
Is there a sideways thumb emoji? In the context of osu! we'll probably not be using this, but I don't know how useful it'll be for other projects. @default0 would you use this in your game?

**Comment by VPeruS** [comment]
Why not to rely on external parser, if there are any?

**Comment by smoogipoo** [comment]
@VPeruS by external parser do you mean a markdown library, or one coded specifically by other projects?

osu! will use a markdown parser in the end, but other projects might find it useful to have this as a basis for simple text formatting where full markdown formatting might not be necessary.

**Comment by smoogipoo** [comment]
Given that there has been almost a week with no more votes on @peppy 's post, I think I'm keen to get this in if he agrees with it. I believe we may even want this in chat, even.

**Comment by peppy** [comment]
i dunno, @Tom94 ?

**Comment by VPeruS** [comment]
Something like this http://www.antlr.org/index.html but more lightweight

**Comment by default0** [comment]
@VPeruS I assume you mean a set of regexes to specify and a callback to apply custom formatting to everything matched by the given regex? Thus allowing users to specify regexes like \*.+\*? Sounds like a more solid approach to me, if that is what you are getting at.

**Comment by VPeruS** [comment]
@default0 kind of.

**Comment by Tom94** [comment]
I'm fine with this going in. At least in my original implementation (which this looks like it uses the core of) I did test the corner cases @default0 mentioned in his earlier comment, and they gave reasonable resultes.

**Comment by default0** [comment]
My reasons for disliking the direction of this is because the whole approach seems rather heavy-handed to me. The base class is essentially useless without an overridden class, which seems odd given that it feels like I should be able to specify at least _some_ form of custom formatting off-the-shelf.
The requirement and simultaneously ability to specify custom data associated with markers also adds to the necessary boiler-plate that using this API incurs.
And last but not least, from what I can tell this only supports delimiters, anything else (say, custom formatting for specific words or phrases) cannot be done with this class.

Contrasting this with the approach @VPeruS outlined, I'm massively in favor of his idea over this proposal.

**Comment by smoogipoo** [comment]
@default0 Formatting is very font-dependent, I don't see it as a bad thing being barebones like this given that. Code structure wise it could do with some improvements, but I'm thinking simplicity-wise this is in a good position.

**Comment by default0** [comment]
@smoogipoo my reason for complaint is not that you have to write the formatting code yourself (ie the code that actually changes text color, makes stuff bold or whatever else). My reason for complaint is the boiler-plate that is necessary from the outside to get to the point of writing the code that actually matters (ie the code that actually changes text color etc). And this is also what I meant when I stated it is heavy-handed.

And yes, while the concrete changes you may want to code to get the text to look how you want is very font dependent, the code to "select" which parts of text should be formatted is a _lot_ less flexible than could be achieved even by easier parsing implementations (fe the list of regexes proposed by @VPeruS).

**Comment by peppy** [comment]
For the time being I'm going to close this one. We can reopen it in the future if a use case comes up.
